### PR TITLE
Auditing using Mojo events

### PR DESCRIPTION
**Pull request by aaannz**
This is proof-of-concept for auditing using mojo events.
Idea: when something happen, event is fired and audit log will write down what happened.

I added multiple event emitters throughout the code (3dfbd2c). These emitter use emit_event helper which adds current user id and connection id to the event (0fd9246).
So far, for demonstration purposes, audit plugin logs only to audit log file (08055f5), but proposed DB schema is included (82c84f1) and db migration files are not yet included.
I have an idea to link those events to event source, if applicable, but to avoid cluttering audit db I propose to update other db tables with event id field which can than point to audit db to see who and when did what.

I needed to stash current user even when going through API access (551661b). I don't think it has any negative side effects, but please check.

### CODE REVIEW

File: lib/OpenQA/Schema/Result/AuditLog.pm [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package OpenQA::Schema::Result::AuditLog;
+use strict;
+use base qw/DBIx::Class::Core/;
+
+# use db_helpers;
+
+__PACKAGE__->table('audit_log');
```
Comment (coolo) [inline lib/OpenQA/Schema/Result/AuditLog.pm]: we use plural for all other result classes - and tables. I admit that logs sounds weired, but it's still inconsistent. Perhaps AuditLogEntries?

File: public/javascripts/audit_log.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+function setup_audit_table()
+{
```
Comment (coolo) [inline public/javascripts/audit_log.js]: as Ancor protested so loudly last time he actually reviewed - we're moving towards proper javascript names. Which are camel case

File: lib/OpenQA/WebAPI.pm [hunk 3]
```diff
@@ -21,6 +21,7 @@ use OpenQA::Schema;
 use OpenQA::WebAPI::Plugin::Helpers;
 use OpenQA::IPC;
 use OpenQA::Worker::Common qw/ASSET_DIR/;
+use Data::Dump qw/pp/;
```
Comment (coolo) [inline lib/OpenQA/WebAPI.pm]: this looks like a left over - take care that @okurz doesn't get you!

**Comment by coolo** [comment]
I'm missing needle creation as event and I wonder how a restart would end up audited. Is that job_set_command?

File: public/javascripts/audit_log.js (hunk 2 above)
Comment (aaannz) [inline public/javascripts/audit_log.js]: Yeah, this js code is not finished anyway.I'll keep that in mind however.

**Comment by aaannz** [comment]
Right, I forgot about needles! ><
About the restart, I must admit I didn't try it with running worker yet. I was playing only with admin interface. I have similar code for Scheduler and WebSockets (as was in old PR from summer) with more detailed info about what's going on there. Anyway, I'm trying it now and lets see.

**Comment by aaannz** [comment]
Yesterday I added 4d01bfd (originally different hash. Well all is rebased now so hashes in description no longer applies), that changed all events to have prefix `openqa_`. This is to differentiate from other events for external plugins like fedmsg and from future `openqa_scheduler` and `openqa_websockets`.
Also added events for needle deletion and create/update and when comment is added.
Today I improved it a little and log only successful events, thus I removed _req and _res suffices. Errors should be visible from regular log, no need to advertise them IMO.

I also tried to standardize on what events` data, but as it is such a diverse set I only changed that events does have id of related event, whatever that means - when comment is added, its id of that comment, when job is duplicated its the id of original job and result is id of new job, etc.

If there are no more comments on the architecture, I'll add db migration data and proper viewer fetching and interlinking data from db.

**Comment by aaannz** [comment]
Rebased, reorganized and file based audit log removed.
Now here is DB based audit log with proper admin viewer.

**Comment by aaannz** [comment]
This is now ready to be reviewed and merged since it is in usable state and I have some other work based on this already.
What I would like to improve next (besides add missing events if found) is to add easier filtering for events. Some initial idea is part of this PR but is not activated. And then displaying of individual events based on GET param.

**Pull request merged**